### PR TITLE
New feature: NATS bindings

### PR DESCRIPTION
**Pull request by jakubnoga**

### CODE REVIEW

**Comment by Lazzaretti** [comment]
With that PR, we can also update the readme and tick the box "NATS Protocol Binding" for v1.0, right?

File: src/binding/nats/mod.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+//! This module provides bindings between [cloudevents-sdk](https://docs.rs/cloudevents-sdk) and [nats](https://docs.rs/nats)
+//! ## Examples
+//! Deserialize [nats::Message](https://docs.rs/nats/latest/nats/struct.Message.html) into [Event](https://docs.rs/cloudevents-sdk/latest/cloudevents/event/struct.Event.html)
+//! ```rust
```
Comment (Lazzaretti) [inline src/binding/nats/mod.rs:4]:
```suggestion
//! ```
```
it's not necessary to set the language to rust; it's the default.
In addition, it is consistant compared to all other crates.

File: src/binding/nats/mod.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+//!
+//!       let event = EventBuilderV10::new()
+//!           .id("123".to_string())
+//!           .ty("example.test")
+//!           .source("http://localhost/")
+//!           .data("application/json", json!({"hello": "world"}))
+//!           .build()
+//!           .unwrap();
+//!
+//!       nc.publish("whatever.subject.you.like", NatsCloudEvent::from_event(event).unwrap());
```
Comment (Lazzaretti) [inline src/binding/nats/mod.rs:36]:
```suggestion
//!       nc.publish("whatever.subject.you.like", NatsCloudEvent::from_event(event).unwrap()).unwrap();
```

File: example-projects/nats-example/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+use std::{error::Error, thread};
+
+use cloudevents::binding::nats::{MessageExt, NatsCloudEvent};
+use cloudevents::{EventBuilder, EventBuilderV10, Event};
+use serde_json::json;
+
+/// First spin up a nats server i.e. 
+/// ```bash
+/// docker run -p 4222:4222 -ti nats:latest
```
Comment (Lazzaretti) [inline example-projects/nats-example/src/main.rs:9]: is `latest` a good idea or should we point to an exact version here?

File: src/binding/nats/mod.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+//! This module provides bindings between [cloudevents-sdk](https://docs.rs/cloudevents-sdk) and [nats](https://docs.rs/nats)
+//! ## Examples
+//! Deserialize [nats::Message](https://docs.rs/nats/latest/nats/struct.Message.html) into [Event](https://docs.rs/cloudevents-sdk/latest/cloudevents/event/struct.Event.html)
```
Comment (Lazzaretti) [inline src/binding/nats/mod.rs:3]: is `latest` a good idea or should we point to an exact version here?

File: example-projects/nats-example/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[package]
```
Comment (Lazzaretti) [inline example-projects/nats-example/Cargo.toml:1]: we should probably also add a build step to `.github/workflows/rust_tests.yml`

**Review by Lazzaretti — CHANGES_REQUESTED** [review]

File: example-projects/nats-example/src/main.rs (hunk 3 above)
Comment (jakubnoga) [reply example-projects/nats-example/src/main.rs:9]: I got that from [Nats Quick Start](https://docs.nats.io/running-a-nats-service/introduction/installation) and I guess it's safe to run these example code with latest version of Nats Server.

File: src/binding/nats/mod.rs (hunk 4 above)
Comment (jakubnoga) [reply src/binding/nats/mod.rs]: Good point, I changed it to `0.21.0`

File: src/binding/nats/deserializer.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            None,
+        );
+
+        let actual = nats_message.to_event().unwrap();
+
+        assert_eq!(expected, actual)
+    }
+
+    #[test]
+    fn test_structured_deserialize_v03() {
```
Comment (jakubnoga) [inline src/binding/nats/deserializer.rs:63]: I added another test and put both v0.3 and v1.0 ticks in the readme.md

File: example-projects/nats-example/src/main.rs (hunk 3 above)
Comment (Jarema) [reply example-projects/nats-example/src/main.rs:9]: I think it's fine to stick to the `latest`.

We usually advise our users to be up to date with latest released `nats-server` version.

Even in the case of NATS 3.0 anywhere in the future, I assume this version of the client will bind to the current protocol compatibility version when talking to the server.

**Comment by Jarema** [comment]
Thanks @jakubnoga ! This PR will make it easier to work with NATS when leveraging Cloud Events.

One note: There is also `async-nats` crate. Might be good to add that position to the protocol binding list (as unchecked) and check it out soon, hopefully, with future PR.

**Review by Lazzaretti — APPROVED** [review]

**Review by Lazzaretti — CHANGES_REQUESTED** [review]
Loos good!
One last thing: Please run `cargo fmt`. It will change 2 lines in the example.

**Comment by jakubnoga** [comment]
> Loos good! One last thing: Please run `cargo fmt`. It will change 2 lines in the example.

@Lazzaretti I did run it before, and once again today just to be sure (both `cargo fmt` and ` cargo fix --all && cargo fmt --all`) and nothing changes. Do you have any clue what could be wrong with my local env?

**Comment by Lazzaretti** [comment]
> > Loos good! One last thing: Please run `cargo fmt`. It will change 2 lines in the example.
> 
> @Lazzaretti I did run it before, and once again today just to be sure (both `cargo fmt` and ` cargo fix --all && cargo fmt --all`) and nothing changes. Do you have any clue what could be wrong with my local env?

There was a bug with rustup, may try `rustup update`?
https://github.com/rust-lang/rustup/issues/2737

To fix it, for now, I pushed the correct formatting.

**Review by Lazzaretti — APPROVED** [review]

**Pull request merged**